enterprise/internal/campaigns/service_apply_campaign_test.go

```diff
@@ -669,8 +850,88 @@ func TestServiceApplyCampaign(t *testing.T) {
 
 				// Assert that the changeset has been updated to point to the new spec
 				assertions.currentSpec = spec2.ID
+				// Assert that the previous spec is still spec 1
+				assertions.previousSpec = spec1.ID
+				assertions.reconcilerState = campaigns.ReconcilerStateQueued
+				assertChangeset(t, attachedChangeset, assertions)
+			})
+
+			t.Run("changeset has failed closing before re-attaching", func(t *testing.T) {
+				campaignSpec1 := createCampaignSpec(t, ctx, store, "detach-reattach-failed-changeset", admin.ID)
+
+				specOpts := testSpecOpts{
+					user:         admin.ID,
+					repo:         repos[0].ID,
+					campaignSpec: campaignSpec1.ID,
+					headRef:      "refs/heads/detached-reattach-failed",
+				}
+				spec1 := createChangesetSpec(t, ctx, store, specOpts)
+
+				// STEP 1: We apply the spec and expect 1 changeset.
+				campaign, changesets := applyAndListChangesets(adminCtx, t, svc, campaignSpec1.RandID, 1)
+
+				// Now we update the changeset so it looks like it's been published
+				// on the code host.
+				c := changesets[0]
+				setChangesetPublished(t, ctx, store, c, "80022", specOpts.headRef)
+
+				assertions := changesetAssertions{
+					repo:             c.RepoID,
+					currentSpec:      spec1.ID,
+					externalID:       c.ExternalID,
+					externalBranch:   c.ExternalBranch,
+					externalState:    campaigns.ChangesetExternalStateOpen,
+					ownedByCampaign:  campaign.ID,
+					reconcilerState:  campaigns.ReconcilerStateCompleted,
+					publicationState: campaigns.ChangesetPublicationStatePublished,
+					diffStat:         testChangsetSpecDiffStat,
+				}
+				reloadAndAssertChangeset(t, ctx, store, c, assertions)
+
+				// STEP 2: Now we apply a new spec without any changesets.
+				campaignSpec2 := createCampaignSpec(t, ctx, store, "detach-reattach-failed-changeset", admin.ID)
+				applyAndListChangesets(adminCtx, t, svc, campaignSpec2.RandID, 0)
+
+				// Our previously published changeset should be marked as "to be closed"
+				assertions.closing = true
+				assertions.reconcilerState = campaigns.ReconcilerStateQueued
+				// And the previous spec is recorded.
+				assertions.previousSpec = spec1.ID
+				c = reloadAndAssertChangeset(t, ctx, store, c, assertions)
+
+				// Now we update the changeset to fail the close task.
+				setChangesetFailed(t, ctx, store, c)
+				assertions.closing = true
+				assertions.reconcilerState = campaigns.ReconcilerStateErrored
+				assertions.externalState = campaigns.ChangesetExternalStateOpen
+
+				// Side-effects of setChangesetFailed.
+				assertions.failureMessage = &canceledChangesetFailureMessage
+				assertions.numFailures = 5
```

Comment:
Since you want to say here that you don't care about the specific values, but need to copy them to make your other assertions work, you can copy them, I think:

```suggestion
				assertions.failureMessage = c.FailureMessage
				assertions.numFailures = c.NumFailures
```